### PR TITLE
Ropes Issue 11 Part 2 R1

### LINKED ISSUE #11: Share Image
#### As a user I want to share an image with someone via email.

__Story__:

This application is called Image Sharer. Sure, you can share images by adding
them to our web service, but how will people who do not know about this web
service be able to see those images? In phone calls with our customers we have
discovered that many of our users are manually emailing their friends links to
the images. That is awesome! We can make things more efficent for our users by
directly supporting sharing images through email.

__Acceptance criteria__:
- [ ] There is a link to share an image.
- [ ] After clicking on this link, a bootstrap modal appears with a form to
  send the email.
- [ ] The form has fields for the recipient's email and an optional message.
- [ ] The email we send has an HTML part and a text part.
- [ ] The email contains the image, the custom message if provided, and a link
  inviting the recipient to check out our application.
- [ ] The modal closes and a success message is shown if the email was sent
  successfully.
- [ ] The user is alerted of errors if the operation was unsuccessful.

__Note__: This story is large, so it makes sense to break the implementation
down into more manageable pieces (a la elephant carpaccio). Please start by
implementing a basic sharing feature that uses a new page (not a modal), and
open a PR for that. Then, once we've verified the simplified version is
working, refactor it to use a modal, and open a second PR for that. The
[video](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project/basics-of-bootstrap-modals-js-and-ajax)
provides some background and code examples that will be really helpful for the
second PR, so it's worth watching (and following along) before you begin that
part.

__Discussion topics__:
- [ ] How could we develop the front end and back end separately? (And why
  might that be a good idea?)
- [ ] How should we implement a non-CRUD action?
- [ ] What are the different types of errors we should handle?
- [ ] What is AJAX? What libraries can help us add AJAX to our Rails app?
- [ ] How should we go about sending emails from our Rails app?
- [ ] How could we show a preview of the image being shared in the modal?
- [ ] What are the implications of allowing users to send emails from our app?

__Reference material__:
- http://guides.rubyonrails.org/action_mailer_basics.html
- https://elements.heroku.com/addons/sendgrid
- https://sendgrid.com/docs/Integrate/Frameworks/rubyonrails.html
- https://github.com/rails/jquery-ujs/wiki
- http://v4-alpha.getbootstrap.com/components/modal
- https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project/basics-of-bootstrap-modals-js-and-ajax

__Dependencies__:
- Image Index
- Refactor flow tests to use page objects

### PR DESCRIPTION
**Pull request by ragurney**
Closes #11 

Supersedes #34

### CODE REVIEW

File: app/assets/javascripts/images/show/share_modal.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import $ from 'jquery';
+
+class ShareModal {
+  constructor(modalSelector){
+    this.$modal = $(modalSelector);
+  }
+
+  attachEventHandlers() {
+    $('#share_modal').on('ajax:error', 'form', function(event, xhr, status, error){
```
Comment (ragurney) [inline app/assets/javascripts/images/show/share_modal.js]: It might be good to include checks for the specific errors you are expecting. This could be done by doing something like: 
```
if (xhr.status === 422)
  ...
if (xhr.status === 404)
  ...
else
  alert('Good god, something has gone terribly wrong!')
```

File: app/views/images/share_form.html.erb [hunk 2]
```diff
@@ -1,9 +0,0 @@
-<%= image_tag @image.url, class: 'share-form--image' %>
```
Comment (ragurney) [inline app/views/images/share_form.html.erb]: You don't have to do it here, but just fyi we usually try to make one commit just for deleting/renaming files, then change the file after that in a separate. This preserves the git blame, rather than git just thinking the entire file is deleted.

File: app/assets/javascripts/images/show/share_modal.js (hunk 1 above)
Comment (ragurney) [inline app/assets/javascripts/images/show/share_modal.js]: In this case I don't think you need a `404` because you're using the show action to render the modal form. But having a catch-all case that just closes the modal and shows a flash message would be nice to have.

File: app/controllers/images_controller.rb [hunk 3]
```diff
@@ -55,6 +52,20 @@ def destroy
...
+    render json: {
+      image: @image,
+      redirect_path: image_url(@image)
+    }, status: :created, location: @image
+  end
+
+  def form_ajax_response_invalid
+    render json: {
+      error: flash[:error],
+      content: (render_to_string partial: 'share_form.html.erb', layout: false)
```
Comment (ragurney) [inline app/controllers/images_controller.rb]: Nice! ✅ 
Do you need the `html.erb` extension here?

File: app/views/images/show.html.erb [hunk 4]
```diff
@@ -10,8 +10,14 @@
...
       <div class="card-block image-card--large__btn-block">
         <%= link_to 'Delete', image_path(@image), class: 'btn btn-danger image-card--large__btn image-card--large__btn--delete', 'data-method': 'delete', 'data-confirm': 'Are you sure you want to delete this image?', rel: 'nofollow' %>
-        <%= link_to 'Share', share_form_image_path(@image), class: 'btn btn-primary image-card--large__btn image-card--large__btn--delete' %>
+        <button type="button" class="btn btn-primary image-card--large__btn" data-toggle="modal" data-target="#share_modal">Share</button>
       </div>
     </div>
   </div>
 </div>
+
+<%= render 'modal' %>
```
Comment (ragurney) [inline app/views/images/show.html.erb]: Hm, I'm not sure how I feel about having the show action also render the modal. I guess it's fine here because we only have one image. Another approach would have been to make a separate action that returns the modal _body_, and then make a `get` request to that when someone clicks the share button.

File: test/page_objects/images/share_modal.rb [hunk 5]
```diff
@@ -1,19 +1,24 @@
 module PageObjects
   module Images
-    class ShareFormPage < PageObjects::Document
-      path :share_form_image
-      path :send_email_image # failed create
+    class ShareModal < AePageObjects::Element
+      # path :image
```
Comment (ragurney) [inline test/page_objects/images/share_modal.rb]: Do you still need these?

File: test/page_objects/images/show_page.rb [hunk 6]
```diff
@@ -20,7 +33,7 @@ def delete
 
       def share
         node.click_on('Share')
-        window.change_to(ShareFormPage)
+        # window.change_to(ShareFormPage)
```
Comment (ragurney) [inline test/page_objects/images/show_page.rb]: ?

File: app/assets/javascripts/images/show/share_modal.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+      $(this).replaceWith(xhr.responseJSON.content);
```
Comment (ragurney) [inline app/assets/javascripts/images/show/share_modal.js]: Ah, now I see why you needed the `form` param. I think it might be clearer to set an instance variable `this.$shareForm = $('#new_share_form');` and then replace the html like this: `this.$shareForm.html(xhr.responseJSON.content)`, but that's just a matter of preference.

File: app/assets/javascripts/application.js [hunk 8]
```diff
@@ -12,3 +12,7 @@
 //
 //= require jquery
 //= require jquery_ujs
+//= require tether
+//= require bootstrap
```
Comment (gitlev) [inline app/assets/javascripts/application.js]: I think bootstrap has to be included after jquery and tether before bootstrap

File: app/assets/javascripts/images/show/main.js [hunk 9]
```diff
@@ -0,0 +1,9 @@
+import $ from 'jquery';
+
+import ShareModal from './share_modal';
+const shareModal = new ShareModal('#share_form');
+shareModal.attachEventHandlers();
+
```
Comment (gitlev) [inline app/assets/javascripts/images/show/main.js]: Thanks

File: app/assets/javascripts/images/show/share_modal.js (hunk 1 above)
Comment (gitlev) [inline app/assets/javascripts/images/show/share_modal.js]: tried it, but it didn't work, my understanding is this function allows for dynamic binding of form elements that are children of share_modal, this way when my ajax:error replaces the form with a new form, it will still be bound to this handler

Comment (gitlev) [inline app/assets/javascripts/images/show/share_modal.js]: :+1:

File: app/assets/javascripts/images/show/share_modal.js (hunk 7 above)
Comment (gitlev) [inline app/assets/javascripts/images/show/share_modal.js]: :+1:

File: app/views/images/share_form.html.erb (hunk 2 above)
Comment (gitlev) [inline app/views/images/share_form.html.erb]: how do you usually handle this?
are you typically committing immediately on creating a file? or do you make new commits on rebase edit?

File: app/controllers/images_controller.rb (hunk 3 above)
Comment (gitlev) [inline app/controllers/images_controller.rb]: without the extension I get this error...
image

Comment (gitlev) [inline app/controllers/images_controller.rb]: nvm fixed

File: test/page_objects/images/share_modal.rb (hunk 5 above)
Comment (gitlev) [inline test/page_objects/images/share_modal.rb]: not sure what your referring to

File: app/views/images/share_form.html.erb (hunk 2 above)
Comment (ragurney) [inline app/views/images/share_form.html.erb]: Typically you make a commit just for the rename of the file, then edit it. You _could_ do this in a rebase by rebasing back to the commit that is before the changes made to the file, then rename the file, make that its own commit, then continue the rebase. Git should be able to figure out the history from there.
